sql/src/main/java/io/crate/metadata/doc/DocTableInfo.java

```diff
@@ -197,42 +184,47 @@ private GroupShardsIterator getShardIterators(WhereClause whereClause,
         );
     }
 
-    private Routing getRouting(final ClusterStateObserver observer,
-                               final WhereClause whereClause,
-                               @Nullable final String preference,
-                               final int currentRetry,
-                               final boolean retry) {
-        ClusterState clusterState = observer.observedState();
-        final Map<String, Map<String, List<Integer>>> locations = new TreeMap<>();
+    @Override
+    public Routing getRouting(final WhereClause whereClause, @Nullable final String preference) {
+        ClusterStateObserver observer = new ClusterStateObserver(clusterService, ROUTING_FETCH_TIMEOUT, logger);
+        final SettableFuture<Routing> routingSettableFuture = SettableFuture.create();
+        observer.waitForNextChange(
```

Comment:
would probably make sense to use the observer only if there are some shards missing?